Do not artificially separate message handler and uit

tautschnig opened the pull request:
They were passed around independently, even though they originated from the very
same object.

tautschnig committed:
Do not artificially separate message handler and uit

They were passed around independently, even though they originated from the very
same object.

allredj reviewed (commented):
This PR failed Diffblue compatibility checks (cbmc commit: 31e6b52c2403e123b70919ca5c3c9bf0d77740ae).
Build URL: https://travis-ci.com/diffblue/test-gen/builds/77146181
Status will be re-evaluated on next push.
Please contact @peterschrammel, @thk123, or @allredj for support.

Common spurious failures:
  - the cbmc commit has disappeared in the mean time (e.g. in a force-push)
  - the author is not in the list of contributors (e.g. first-time contributors).
##

tautschnig commented:
FWIW: this is an API change, and will thus a need for adjusting TG is expected. Or at least that's what I hope the reported TG failure is about.

peterschrammel commented:
@thk123, can you create a TG bump for this, please?

thk123 commented:
Started - though could use some guidance on exactly how to apply this.  For example, we have:

```C++
cbmc_solverst cbmc_solvers(bmc_options, symbol_table, sub_solver_out);
```

`sub_solver_out` is either a `message_handlert` or a `null_message_handlert` (decided by a define) and not entirely trivial for me to change either to be a `ui_message_handlert`. What is the purpose of this, why does `cbmc_solverst` have a tight dependency on what output to use?

tautschnig commented:
> What is the purpose of this, why does `cbmc_solverst` have a tight dependency on what output to use?

~That's a fair question, and I've got another commit in #2310 (f578d33) that removes the unused `ui` information from string refinement.~(done in #2542) What isn't done yet is completely removing it from refinement, but that would be a functional change, not only cleanup.

With regard to guidance: In your current set-up, somehow you must be picking a ui despite just having a `message_handlert` or `null_message_handlert`. So whatever selects this ui should be investigated, because it problably has a `ui_message_handlert`. If it doesn't, then you've genuinely decoupled the message handler from the ui configuration, which seems broken to me.

tautschnig commented:
@kroening @peterschrammel Could you please take another look at this one, especially given I added the commit `Revert "output the iteration number in XML"`.

Overall, options/configuration processing in refinement seems a bit messy I must say. @romainbrenguier would you mind taking a look at that? Notably, `string_refinementt` shadows `config_` with a member of a different type, with some of its options, however, looking rather similar. It's not completely related to this PR, but maybe a cleanup of this code can be taken on someone's agenda.

peterschrammel commented:
> "output the iteration number in XML"

This is a feature required by a customer. It cannot just be reverted, but must be fixed in some other way.

tautschnig committed:
Change uit configuration to xml-specific Boolean choice

tautschnig commented:
> It cannot just be reverted, but must be fixed in some other way.

Thanks a lot for the info. I've left this in place and went for a path that should easily be fixable in TG (or any other user of `cbmc_solverst`) as well. @thk123 Could you please look into this?

peterschrammel commented:
Let's go with this. It's certainly an improvement over the status quo.

tautschnig commented:
I believe this still requires a proper TG bump, or at least that's what I'd infer from @thk123's earlier comments. Thus handing over to @thk123 (or anyone else who is able to look into this).

thk123 reviewed (dismissed):
Currently doesn't work with TG

I need to be able to construct a `bmct` and `cbmc_solverst` with a `null_message_handlert` (that is - I want the output to be binned). `cbmc_solverst` can be done by passingin a `false` for `_output_xml_in_refinement`, but I can't see a way to do it for `bcmt`

tautschnig commented:
> Currently doesn't work with TG
> 
> I need to be able to construct a `bmct` and `cbmc_solverst` with a `null_message_handlert` (that is - I want the output to be binned). `cbmc_solverst` can be done by passingin a `false` for `_output_xml_in_refinement`, but I can't see a way to do it for `bcmt`

I will hack around this, but I'd appreciate if @thk123 and @peterschrammel could get together about this. Essentially, the setup here is broken as using a `bmct` without a `ui_message_handlert` is just broken, and the Boolean `output_xml_in_refinement` is a terrible hack. It's equally broken that a `ui_message_handlert` doesn't actually accept an underlying `message_handlert`, but instead has hardcoded message handlers.

tautschnig commented:
I have taken another look at this. Using a `bmct` without a `ui_message_handlert` is just wrong. There are several places where `bmct` generates UI-dependent output.

IMHO, the correct path forward is for `ui_message_handlert` to be parameterised by a `message_handlert`, which should resolve the problem. I'll try to sort this out.

tautschnig committed:
Restrict get_json_stream to ui_message_handler

... now that we properly pass around ui_message_handlert we can clean up the
JSON stream handling as well.

tautschnig commented:
> IMHO, the correct path forward is for `ui_message_handlert` to be parameterised by a `message_handlert`, which should resolve the problem. I'll try to sort this out.

Done. One can now create a `ui_message_handlert` using, e.g., a `null_message_handlert`. @thk123 could you please take another look whether this makes things viable for TG? Also, @peterschrammel @NlightNFotis would you mind taking a look? I do recall a long debate on #1814, and the changes in the first commits of this PR enabled what I perceive as a cleanup. Equally, those changes might break TG in some way and may need changes similar to those applied in the commits here.

Edit/note: reviewing commit-by-commit is advisable.

peterschrammel commented:
Rebased and TG bump created.

allredj reviewed (commented):
This PR failed Diffblue compatibility checks (cbmc commit: eefad584e4e70d765e75d6a1107329a75cf9a36e).
Status will be re-evaluated on next push.
Please contact @peterschrammel, @thk123, or @allredj for support.

Common spurious failures:
  - the cbmc commit has disappeared in the mean time (e.g. in a force-push)
  - the author is not in the list of contributors (e.g. first-time contributors).

allredj reviewed (commented):
This PR failed Diffblue compatibility checks (cbmc commit: 8eb05f71283771499cbf7f2b394880fc6cea98f7).
Build URL: https://travis-ci.com/diffblue/test-gen/builds/84891985
Status will be re-evaluated on next push.
Please contact @peterschrammel, @thk123, or @allredj for support.

Common spurious failures:
  - the cbmc commit has disappeared in the mean time (e.g. in a force-push)
  - the author is not in the list of contributors (e.g. first-time contributors).

peterschrammel commented:
@kroening, that's ready to merge.

kroening commented:
I would need a bit more background on this one. What makes me nervous is that there's now ui_message_handler inside goto-cc, which doesn't intend to do any xml or json output, or interact with any UI.

tautschnig commented:
> What makes me nervous is that there's now ui_message_handler inside goto-cc, which doesn't intend to do any xml or json output, or interact with any UI.

The console equally is a user interface, isn't it?

peterschrammel commented:
`ui_message_handlert` was already inside goto-cc through `gcc_message_handlert` and `console_message_handlert`.

kroening commented:
Ah, we are anyway trying to get rid of language_uit.

peterschrammel commented:
> Ah, we are anyway trying to get rid of language_uit.

Is this an 'approve'?

Edit: https://github.com/diffblue/cbmc/pull/2980 needs to be merged first, then rebase.

kroening commented on src/solvers/refinement/string_refinement_util.h:
```diff
@@ -9,6 +9,8 @@
 
+#include <memory>
+
 #include "string_builtin_function.h"
```
The new include seems to be spurious?

tautschnig replied:
I'm afraid it isn't, this file uses `std::unique_ptr`, which previously worked as it was transitively included via message handler includes.

kroening commented:
After rebasing to #2980, the changes in goto-cc should no longer be necessary. The rest is fine!

peterschrammel committed:
Make ui_message_handlert parameterizable in the underlying message handler

allredj reviewed (commented):
This PR failed Diffblue compatibility checks (cbmc commit: 0c8b10880f63bc1d709b912f23a60caff4349a0f).
Build URL: https://travis-ci.com/diffblue/test-gen/builds/85486179
Status will be re-evaluated on next push.
Please contact @peterschrammel, @thk123, or @allredj for support.

Common spurious failures:
  - the cbmc commit has disappeared in the mean time (e.g. in a force-push)
  - the author is not in the list of contributors (e.g. first-time contributors).
